carp.common/src/commonMain/kotlin/dk/cachet/carp/common/infrastructure/services/ApplicationServiceRequestLogger.kt

```diff
@@ -0,0 +1,46 @@
+package dk.cachet.carp.common.infrastructure.services
+
+import dk.cachet.carp.common.application.services.ApplicationService
+import dk.cachet.carp.common.application.services.IntegrationEvent
+
+
+/**
+ * Notifies of incoming requests and responses through [log],
+ * as well as events preceding the request and fired as a result of the request.
+ */
+class ApplicationServiceRequestLogger<
+    TService : ApplicationService<TService, TEvent>,
+    TEvent : IntegrationEvent<TService>,
+    TRequest : ApplicationServiceRequest<TService, *>
+>(
+    private val eventBusLog: EventBusLog,
+    private val log: (LoggedRequest<TService, TEvent>) -> Unit = { },
+    val decoratee: Command<TRequest>
+) : Command<TRequest>
+{
+    override suspend fun <TReturn> invoke( request: TRequest ): TReturn
+    {
+        @Suppress( "UNCHECKED_CAST" )
+        fun getCurrentEvents() = eventBusLog.retrieveAndEmptyLog() as List<TEvent>
```

Comment:
`EventBusLog` logs _any_ `IntegrationEvent<TService>`. Thus, events coming from multiple services.

But, now I'm wondering whether this is even correct then. I imagine it'll also include other `TEvent`'s. So I think you caught something misleading there, but which doesn't break at runtime since the JVM has type erasure either way.

Reply:
I removed the erroneous/unchecked cast, and realized `TEvent` in fact can't be guaranteed for the most part here. Removing it leads to more readable code, and the code operating on these logged events only deal with `IntegrationEvent<*>` either way.

Fixed in a new commit!